drgn_tools/memstate.py

```diff
@@ -0,0 +1,846 @@
+    print_pretty_gb_l2("Shared mem", userspace_mm["sh_mm"])
+
+
+def kernel_mm(prog, mm_stats, mem_total, mem_free, user_alloc):
+    """computing the kernel memory usage statistics and return it as dictionnary"""
+    page_size = prog["PAGE_SIZE"].value_()
+    slab_mm = pages_to_gb(page_size, mm_stats["Slab"])
+    vmalloc = pages_to_gb(page_size, mm_stats["VmallocUsed"])
+    page_tables = pages_to_gb(page_size, mm_stats["PageTables"])
+    kernel_stack = pages_to_gb(page_size, mm_stats["KernelStack"])
+    percpu = pages_to_gb(page_size, mm_stats["Percpu"])
+    total = slab_mm + vmalloc + page_tables + kernel_stack + percpu
+    hugepagessize = get_total_hugetlb_pages(prog)
+    ukn = mem_total - (mem_free + total + user_alloc + hugepagessize)
+    total = round(total + ukn, 1)
+    return {
+        "slab_mm": slab_mm,
+        "vmalloc": vmalloc,
+        "page_tables": page_tables,
+        "kernel_stack": kernel_stack,
+        "percpu": percpu,
+        "total": total,
+        "ukn": ukn,
+    }
+
+
+def display_kernel_mm(kernel_mm):
+    """
+    displaying the userspace memory usage , passing the userspace_mm calculated before as parameter
+    """
+    print_pretty_gb_l1("Kernel", round(kernel_mm["total"], 1))
+    print_pretty_gb_l2("Slabs", round(kernel_mm["slab_mm"], 1))
+    print_pretty_gb_l2("Percpu", round(kernel_mm["percpu"], 1))
+    print_pretty_gb_l2("Unknown", round(kernel_mm["ukn"], 1))
+
+
+def memory_usage(prog: Program, hp_stats):
+    """ " This is a wrapper API that displays a memory summary"""
+    print("")
+    pages_size = prog["PAGE_SIZE"].value_()
+    print("MEMORY USAGE SUMMARY (in GB):")
+    mm_stats = get_all_meminfo(prog)
+    total_memory = pages_to_gb(pages_size, totalram_pages(prog).value_())
+    free_memory = pages_to_gb(pages_size, mm_stats["MemFree"])
+    used_memory = total_memory - free_memory
+
+    print_pretty_gb("Total memory", round(total_memory, 1))
+    print_pretty_gb("Free memory", round(free_memory, 1))
+    print_pretty_gb("Used memory", round(used_memory, 1))
+    user_space_memory = user_space_mm(prog, mm_stats)
+    display_userspace_mm(user_space_memory)
+    kernel_mm_usage = kernel_mm(
+        prog, mm_stats, total_memory, free_memory, user_space_memory["total"]
+    )
+    display_kernel_mm(kernel_mm_usage)
+    display_hugepages_state(prog, hp_stats)
+    used_swap = pages_to_gb(
+        pages_size, mm_stats["SwapTotal"] - mm_stats["SwapFree"]
+    )
+    print_pretty_gb("Swap used", round(used_swap, 1))
+    return {
+        "total_memory": total_memory,
+        "free_memory": free_memory,
+        "used_memory": used_memory,
+        "kernel_mm_usage": kernel_mm_usage,
+        "user_space": user_space_memory,
+    }
+
+
+def get_slab_inof(prog: Program):
+    """
+    this api is returning a dictionnary that contain the slab cache , it size and their aliases
+    """
+    slabs_infos: dict = defaultdict()
```

Comment:
The type annotation `dict` is not usable for python < 3.9. So, we use `typing.Dict` (and `typing.List`, etc).

Further, if you're going to annotate it, you should probably include types for the keys and values. So that would be:

```
from numbers import Number
from typing import Dict, List

slab_infos: Dict[str, Dict[str, Number]] = defaultdict()
```